rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java

```diff
@@ -0,0 +1,209 @@
+                break;
+            default:
+                throw new InvalidLRAParticipantDefinitionException(
+                    method.toGenericString() + ": invalid number of arguments: " + method.getParameterCount());
+        }
+
+        Response.ResponseBuilder builder = Response.status(Response.Status.OK);
+
+        if (result instanceof CompletionStage) {
+            // TODO optimize this with coordinator
+            // wait for result the same way as JAX-RS filters do
+            try {
+                result = ((CompletionStage) result).toCompletableFuture().get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (result == null) {
+            builder.status(Response.Status.NOT_FOUND);
+        } else if (method.getReturnType().equals(Void.TYPE)) {
+            // void return type and no exception was thrown
+            builder.entity(type.equals(COMPLETE) ? ParticipantStatus.Completed.name() : ParticipantStatus.Compensated.name());
+        } else if (result instanceof ParticipantStatus) {
+            builder.entity(((ParticipantStatus) result).name());
+        } else if (result instanceof Response) {
+            return (Response) result;
+        } else {
+            throw new InvalidLRAParticipantDefinitionException(
+                method.toGenericString() + ": invalid type of returned object: " + result.getClass());
+        }
+
+        return builder.build();
+    }
+
+    private Object invokeMethod(String type, Method method, Object o, Object... args) {
+        try {
+            return method.invoke(o, args);
+        } catch (InvocationTargetException e) {
+            if (e.getTargetException() instanceof WebApplicationException) {
+                return ((WebApplicationException) e.getTargetException()).getResponse();
+            }
+
+            // other exceptions differ based on the participant method type
+            if (type.equals(COMPENSATE)) {
+                return ParticipantStatus.FailedToCompensate.name();
+            } else if (type.equals(COMPLETE)) {
+                return ParticipantStatus.FailedToComplete.name();
+            } else {
+                // @Status and @Forget should return HTTP 500
+                return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+            }
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    void setCompensateMethod(Method compensateMethod) {
+        this.compensateMethod = compensateMethod;
+    }
+
+    void setCompleteMethod(Method completeMethod) {
+        this.completeMethod = completeMethod;
+    }
+
+    void setStatusMethod(Method statusMethod) {
+        this.statusMethod = statusMethod;
+    }
+
+    void setForgetMethod(Method forgetMethod) {
+        this.forgetMethod = forgetMethod;
+    }
+
+    Method getCompensateMethod() {
```

Comment:
This is the only getter method that you need since if the compensate method is absent then by definition it is not a participant (see my earlier comment about how to identify a participant resource), ie the other getters can be removed.

Reply:
A *participant must define a compensate method* and all of the other methods are optional. Therefore, if this method is present you need to process the java class and if it is not present you should ignore the class (even if it contains any of the optional annotations since mixing and matching with JAX-RS annotated methods will be a recipe for disaster).